src/cljam/util/intervals.clj

```diff
@@ -0,0 +1,26 @@
+(ns cljam.util.intervals
+  "Data structures for interval queries.")
+
+(defprotocol IIntervals
+  (find-overlap-intervals [this chr start end]
+    "Finds intervals that is overlap [start end]."))
+
+(deftype Sorted-map-intervals [^clojure.lang.Sorted smap]
+  IIntervals
+  (find-overlap-intervals [this chr start end]
+    (mapcat (fn [[_ ls]] (drop-while #(> start (:end %)) ls))
+            (subseq (get (.smap this) chr) <= end))))
+
+(defn make-sorted-map-intervals
+  "Returns Sorted-map-intervals that has a sorted-map (start is key)
+   for eaach chr calculated from intervals."
+  [intervals]
+  (->> (group-by :chr intervals)
+       (into {}
+             (map
+              (fn [[chr xs]]
+                [chr
+                 (->> (group-by :start xs)
+                      (map (fn [[k vs]] [k (sort-by :end vs)]))
+                      (into (sorted-map)))])))
+       Sorted-map-intervals.))
```

Comment:
How about holding one record for each chromosome like the following?
This enables using different data structures based on a profile of intervals (e.g. no overlaps) in the future.

```clojure
(defprotocol IIntervals
  (find-overlap-intervals* [this start end]))

(deftype SortedMapIntervals [m]
  IIntervals
  (find-overlap-intervals* [this start end]
    (mapcat (fn [[_ x]] (drop-while #(< (:end %) start) x)) (subseq m <= end))))

(defn- make-sorted-map-intervals
  [intervals]
  (->> (group-by :start intervals)
       (map (fn [[k vs]] [k (sort-by :end vs)]))
       (into (sorted-map))
       (->SortedMapIntervals)))

(defn index-intervals
  [intervals]
  (->> intervals
       (group-by :chr)
       (into {} (map (fn [[k vs]] [k (make-sorted-map-intervals vs)])))))

(defn find-overlap-intervals
  [indexed-intervals chr start end]
  (find-overlap-intervals* (get indexed-intervals chr) start end))
```